### PR TITLE
feat: complete the grammar and tidy up the repo

### PR DESCRIPTION
**Pull request by amaanq**
This PR has a lot of changes, so do take your time to look over them. But I hope you'll be pleased with this, as I've fixed several bugs and added a much smoother CI/CD action, as well as general touches making the repo look a lot nicer. I've decided to bump the version to v0.8.0 accordingly, for the actions you can set the secrets yourself if you choose to, just let me know and I can send/email you the CARGO token

### CODE REVIEW

**Comment by amaanq** [comment]
Ping @drom, just hoping you see this! :)

**Comment by drom** [comment]
Thank you for picking up the project @amaanq . LGTM
Let me know when you are ready to merge.

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
```
Comment (drom) [inline .eslintrc.js]: I would prefer keeping eslint rules inside of package.json file

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    branches:
+      - "**"
+
+jobs:
+  test:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: true
+      matrix:
+        os: [macos-latest, ubuntu-latest]
```
Comment (drom) [inline .github/workflows/ci.yml:17]: We should keep 3 separate CI test plans Linux / MacOS / Windows as it was before.
Also run on broad matrix of NodeJS versions.

**Comment by amaanq** [comment]
Sure thing, it's ready imo. For the publish actions, just add your npm token as a secret in the repo, and does emailing you work for my crates secret?

**Review by drom — CHANGES_REQUESTED** [review]
I would prefer a set of more focused PRs
It is rather hard to review so many unrelated changes.
It all looks good to me from 11KM height ;)

File: .github/workflows/ci.yml (hunk 2 above)
Comment (amaanq) [reply .github/workflows/ci.yml:17]: I like the broad matrix of NodeJS versions idea, but what's wrong with consolidating the OS tests to one workflow file?

File: .eslintrc.js (hunk 1 above)
Comment (amaanq) [reply .eslintrc.js]: done

**Comment by amaanq** [comment]
> I would prefer a set of more focused PRs It is rather hard to review so many unrelated changes. It all looks good to me from 11KM height ;)

Sorry about that :sweat_smile: but most of the changes would sort-of depend on each other, with the exception of the manifest/dotfile updates

File: .github/workflows/ci.yml (hunk 2 above)
Comment (drom) [reply .github/workflows/ci.yml:17]: Linux plan usually runs very fast and if the badge is green it make sense to wait for MacOS.
MacOS and Windows plans needed mostly to fight respected OS issues.

**Comment by drom** [comment]
@amaanq BTW i hope you are looking into the latest FIRRTL spec https://github.com/chipsalliance/firrtl-spec/releases/tag/v1.2.0

File: .github/workflows/ci.yml (hunk 2 above)
Comment (amaanq) [reply .github/workflows/ci.yml:17]: > Linux plan usually runs very fast and if the badge is green it make sense to wait for MacOS. MacOS and Windows plans needed mostly to fight respected OS issues.

Well, the mac/windows ones don't take that much longer, and I do take into account differences in the OS (at least for Windows - see the scripts in package.json, Mac is Unix-like so it's ok), see an example of the CI badge here: https://github.com/amaanq/tree-sitter-kdl

**Comment by amaanq** [comment]
> @amaanq BTW i hope you are looking into the latest FIRRTL spec https://github.com/chipsalliance/firrtl-spec/releases/tag/v1.2.0

Is there anything new as far as syntax/file structure goes? I did refer to this when making my changes

**Comment by drom** [comment]
> > @amaanq BTW i hope you are looking into the latest FIRRTL spec https://github.com/chipsalliance/firrtl-spec/releases/tag/v1.2.0
> 
> Is there anything new as far as syntax/file structure goes? I did refer to this when making my changes

There are some syntax changes in master https://github.com/chipsalliance/firrtl-spec/blob/main/revision-history.yaml#L7-L14

**Comment by amaanq** [comment]
Oh wow, quite some changes. Do we want to reflect these changes in the grammar? Like the removal of fixed point types, partial connections, and validif?

**Comment by amaanq** [comment]
The spec pdf still has these removed items, and no mentions of the `const` type qualifier FYI (same with the ANTLR grammar), so idk where that would go

File: .npmignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+/test
```
Comment (drom) [inline .npmignore:1]: instead `.npmignore` I would keep the `files` section inside `package.json` to list what ned to be published to NPM

File: Package.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// swift-tools-version:5.3
```
Comment (drom) [inline Package.swift:1]: What is the purpose of this `swift` file?

File: bin/install.js [hunk 5]
```diff
@@ -1,36 +0,0 @@
-#!/usr/bin/env node
```
Comment (drom) [inline bin/install.js:1]: Why have you deleted `instal` script?

File: bindings/node/binding.cc [hunk 6]
```diff
@@ -1,10 +1,10 @@
+#include "nan.h"
```
Comment (drom) [inline bindings/node/binding.cc:1]: are those just style changes? or functional fixes to?

File: bindings/node/index.js [hunk 7]
```diff
@@ -1,5 +1,3 @@
-'use strict';
```
Comment (drom) [inline bindings/node/index.js:1]: why deleting `use strict` ?

File: bindings/rust/build.rs [hunk 8]
```diff
@@ -1,7 +1,6 @@
 use std::env;
```
Comment (drom) [inline bindings/rust/build.rs]: are those just style changes? or functional fixes to?

File: src/grammar.json [hunk 9]
```diff
@@ -0,0 +1,2518 @@
+{
```
Comment (drom) [inline src/grammar.json:1]: why do we commit generated files to git?

File: src/node-types.json [hunk 10]
```diff
@@ -0,0 +1,1507 @@
+[
```
Comment (drom) [inline src/node-types.json:1]: why do we commit generated files to git?

File: src/parser.c [hunk 11]
```diff
@@ -0,0 +1,17178 @@
+#include <tree_sitter/parser.h>
```
Comment (drom) [inline src/parser.c]: why do we commit generated files to git?
It is not the source code.

File: src/scanner.cc [hunk 12]
```diff
@@ -1,14 +1,15 @@
-#include <tree_sitter/parser.h>
-#include <vector>
-#include <cwctype>
-#include <cstring>
 #include <cassert>
+#include <cstring>
```
Comment (drom) [inline src/scanner.cc]: are those just style changes? or functional fixes to?

**Review by drom — CHANGES_REQUESTED** [review]

File: Package.swift (hunk 4 above)
Comment (amaanq) [reply Package.swift:1]: This adds swift bindings, similar to how we have Rust and Node bindings (it is a common addition in many tree-sitter core repos)

File: bin/install.js (hunk 5 above)
Comment (amaanq) [reply bin/install.js:1]: I don't exactly know when you made this repo, but it seems to be quite a while ago from before tree-sitter handled installation for a repository. If you check out my branch and run npm install, it seems to run almost the same thing as this did, if not anything more/new

File: bindings/node/binding.cc (hunk 6 above)
Comment (amaanq) [reply bindings/node/binding.cc:1]: Style changes from eslint

File: bindings/node/index.js (hunk 7 above)
Comment (amaanq) [reply bindings/node/index.js:1]: I'm not sure of the purpose of this in the context of a tree-sitter grammar file, but it seems to be unnecessary with eslint and the `@ts-check` directive

File: bindings/rust/build.rs (hunk 8 above)
Comment (amaanq) [reply bindings/rust/build.rs]: Style/cleaning up, like removing an unnecessary reference in a function call

File: src/grammar.json (hunk 9 above)
Comment (amaanq) [reply src/grammar.json:1]: It is so those that want to install the parser do not need to generate from the grammar file, which nvim-treesitter  (neovim plugin) does. This way, all one needs to install this parser is a C compiler, which reduces time spent on installing and resources as well, at the cost of having this file in the repo. Diffs and PRs won't show changes to these files because of the .gitattributes file I added if that's a concern of yours

File: src/scanner.cc (hunk 12 above)
Comment (amaanq) [reply src/scanner.cc]: Style (clang format), and I also wanted to add a fuzz tester but it seems that the fuzz action here (https://github.com/vigoux/tree-sitter-fuzz-action) doesn't have cstring/cassert built in (maybe cuz it's running alpine)

File: .npmignore (hunk 3 above)
Comment (amaanq) [reply .npmignore:1]: Hmm gotcha, I do think being a bit more explicit here rather than reading through `package.json` to find the published files is easier on the user/dev, totally your call though

File: bindings/node/binding.cc (hunk 6 above)
Comment (drom) [reply bindings/node/binding.cc:1]: This is C++ file what `eslint` style changes you are talking about?

File: bindings/node/index.js (hunk 7 above)
Comment (drom) [reply bindings/node/index.js:1]: `use strict` is not a comment. It makes JS engine to run code in strict mode. So it in NOT an equivalent to eslint

File: bindings/node/binding.cc (hunk 6 above)
Comment (amaanq) [reply bindings/node/binding.cc:1]: Oh, clang format then oops

File: bindings/node/index.js (hunk 7 above)
Comment (amaanq) [reply bindings/node/index.js:1]: I see, I'll add it back :slightly_smiling_face:

File: package.json [hunk 13]
```diff
@@ -1,55 +1,84 @@
...
       ]
     }
   ],
   "eslintConfig": {
-    "extends": "@drom/eslint-config/eslint4/node8"
+    "env": {
+      "commonjs": true,
+      "es2021": true
+    },
+    "extends": "google",
```
Comment (drom) [inline package.json]: Why are we switching project to `google ` rules?

**Comment by drom** [comment]
> Oh wow, quite some changes. Do we want to reflect these changes in the grammar? Like the removal of fixed point types, partial connections, and validif?

Agree, let's limit this PR to v1.2.0 spec. And make all radical changes when v2.0 will be out.

File: bindings/node/binding.cc (hunk 6 above)
Comment (drom) [reply bindings/node/binding.cc:1]: If we start using `clang` format we should be explicit about the which format we require.

File: src/grammar.json (hunk 9 above)
Comment (drom) [reply src/grammar.json:1]: I am for keeping only source code under source revision control.
Published NPM package will include all generated files.
Same for Cargo and other Package managing systems.
Keeping large generated files under source code revision control leading to confusion and mismatch IMHO

**Review by drom — CHANGES_REQUESTED** [review]

File: package.json (hunk 13 above)
Comment (amaanq) [reply package.json]: Hmm, sorry about that but it's just what I'm used to. Main feature I enjoy is ensuring commas after even the last item.

File: bindings/node/binding.cc (hunk 6 above)
Comment (amaanq) [reply bindings/node/binding.cc:1]: default is llvm style iirc

File: src/grammar.json (hunk 9 above)
Comment (amaanq) [reply src/grammar.json:1]: nvim-treesitter, helix, emacs, nixpkgs, and whatever other editor/pkg management system almost always pulls the git version as oftentimes repos don't publish often enough to fix bugs/add features. I would like to say that the .gitattributes file prevents unnecessary noise in viewing diffs / changes online. See example here: 
image

So, I'd like to counter the idea that this would lead to confusion/mismatches, provided that PRs changing the grammar also update the generated files (ideally with a "generate" commit of sorts)

File: package.json (hunk 13 above)
Comment (drom) [reply package.json]: Well, every project has some rules. let's stick with preexisting project rules for now.

File: src/grammar.json (hunk 9 above)
Comment (drom) [reply src/grammar.json:1]: Git is not a very good "package manager". If some systems can't download / untar NPM package we can create separate branch "package" or "release" or something, similar to "gh-pages" where we will "publish" releases of generated C, H, JSON files (no source) and that would be a "pure-man package manager". How about this?

File: grammar.js [hunk 14]
```diff
@@ -1,337 +1,411 @@
...
+    circuit: $ => seq(
+      'circuit', $.identifier, ':', optional($.info),
+      optional(seq($._indent, repeat($.module), $._dedent)),
     ),
-    seq(
-      'Fixed',
-      optseq('<', $.intLit, '>'),
-      optseq('<', '<', $.intLit, '>', '>')
+
+    module: $ => choice(
```
Comment (drom) [inline grammar.js:66]: @amaanq 
Uh, This is total revamping of content, style, organization, and structure.
Could you summarize factual changes?
What bugs it fixes?
What features where added?
For example, does it fix #2 ?

**Review by drom — CHANGES_REQUESTED** [review]

File: package.json (hunk 13 above)
Comment (amaanq) [reply package.json]: Sure..as I was reverting it back I encountered something unpleasant (which is, as I remember now, why I always use the Google config)

image

However, I can show lsp hover info/invoke a goto action and the function is defined (as noted in my reference types directive on L15). So it seems your config enables "no-undef" which adds a lot of visual noise for anyone using virtual text/underline errors pointing such things out. I disabled this specific rule in the config to remove that

File: grammar.js (hunk 14 above)
Comment (amaanq) [reply grammar.js:66]: Been a while now, but let me summarize it from memory as best as I can

1. mux and other reserved keywords could not be used as identifiers, however in the main repository (firrtl/regress/FPU.fir) mux is used as an identifier, so adding the keyword_identifier rule aliased to identifier and inlining it solves this.

2. better distinguishing of escape sequences

3. add optional const qualifier support in port

4. extract any rules in statements into their own rule, which then enables setting statement as a supertype. supertypes allow us to use them in a query, but hides them in a node - inline doesn't allow us to use them in a query as well.

5. yes, when-else is fixed, see:
image

6. make queries (imo) perfect to neovim treesitter standards (we will standardize this with helix/other editors soon, as Max the tree-sitter creator & others have agreed last-wins is best - aka Neovim convention)

7. use snake_case over camelcase as is convention in tree-sitter grammars

8. ensure every FIRRTL file parses successfully and correctly

9. fill out docs as best as possible in manifests and libraries/packages

10. add swift package

11. format other files like rust, swift header, c/c++ based on appropriate standards

12. update tests & ci as unix-like work together just fine, then windows is its own thing

**Comment by amaanq** [comment]
Just an FYI with the generated files, after this is merged the .gitattributes file will auto hide diffs from those for any future commits/PRs

...and as you can see in the failing CI, your eslint config still has quite some complaints, e.g.
image
image

I updated package.json accordingly for this

File: grammar.js (hunk 14 above)
Comment (amaanq) [reply grammar.js:66]: I also added support to vim/neovim for auto detection of FIRRTL files so anyone who installs the grammar (after I add it to nvim-treesitter after this is merged) will have highlights, code folding, and indents working out of the box :) 

See:
Peek 2023-03-09 21-23

File: src/grammar.json (hunk 9 above)
Comment (amaanq) [reply src/grammar.json:1]: I don't particularly like that as any sources that leverage tree-sitter packages have it configured to either generate from npm if there's no generated files, or just simple compile parser.c if it's present.

Just doing a quick check in nvim-treesitters parsers.lua list, about 30 of the ~200 languages we support require this generation process. So for the most part, grammars contain these files so people on lower end machines won't struggle as much and the only dependency is a C99 compiler then (making it much more cross-platform/universally available)..

Of course it's your final call as it's your repo, but to me it'd have far more benefits than downsides to add them here

File: grammar.js [hunk 15]
```diff
@@ -1,337 +1,411 @@
...
+    sub_field: $ => seq($.expression, '.', $.field_id),
+
+    sub_index: $ => seq($.expression, '[', $.number, ']'),
+
+    sub_access: $ => seq($.expression, '[', $.expression, ']'),
+
+    mux: $ => seq('mux', '(', $.expression, ',', $.expression, ',', $.expression, ')'), // no commas in ANTLR
 
-/* eslint camelcase: 0 */
-/* eslint no-undef: 0 */
```
Comment (drom) [inline grammar.js:336]: `no-undef` and `no-unused-vars` was in this file because they are specific to this file only.

**Review by drom — CHANGES_REQUESTED** [review]
